token_dispenser/token_dispenser_lambda.py

```diff
@@ -0,0 +1,161 @@
+        return False
+    else:
+        return True
+
+def handler(event, context):
+    logging.getLogger().setLevel(logging.INFO)
+    logging.info('Lambda handler received event %s', event)
+    client_id = event.get('client_id', '')
+    minimum_alive_secs: int|None= int(event.get('minimum_alive_secs', config.DEFAULT_TOKEN_MIN_ALIVE_TIME))
+    # client_id must be alphanumeric
+    if not is_client_id_valid(client_id):
+        return {
+            "statusCode": 400,
+            "body": json.dumps({"error": config.ERROR_MISSING_CLIENT_ID})
+        }
+    if not is_minimum_alive_secs_valid(minimum_alive_secs):
+        return {
+            "statusCode": 400,
+            "body": json.dumps({"error": config.ERROR_MINIMUM_ALIVE_SECS})
+        }
+    # if the token expected to be expired shorter than the expiration time, then
+    # get a new token, save new token to dyanoDB with new TTL
+    # minimum_alive_secs:int|None = int(json.loads(event['body'])['minimum_alive_secs'])
+    logger = TokenDispenserLogger.get_instance(client_id=client_id)
+    logger.info(f"client_id with context: {context}")
+    ClientTokens.Meta.table_name = config.DYNAMO_DB_CACHE_TABLE_NAME
+    ClientTokens.Meta.region = config.AWS_REGION
+    logger.info(f"dynamoDB region: {config.AWS_REGION}")
+    logger.info(f"dynamoDB table name: {config.DYNAMO_DB_CACHE_TABLE_NAME}")
+    logger.info("Starting token process")
+    logger.info(f'client_id: {client_id}  minimum_alive_secs: {minimum_alive_secs}')
+
+
+    token_json = None
+    # Retrieve token structure from DynamoDB using client_id
+    try:
+        # When a request comes in, check dynamoDB first, if not found then get new token
+        token_structure_str:str = get_token_by_client_id(client_id)
+        if(token_structure_str is not None):
+            logger.info(f"token structure found in dynamoDB")
+            token_json = json.loads(token_structure_str)
+        else:  # can not find any entry in dynamoDB, create a new one and save to dynamoDB
+            logger.info(f"token structure not found in dynamoDB. Creating new one and saving to dynamoDB.")
+            token_json = get_new_token(client_id)
+            return json.dumps(token_json)
+
+        # This lambda does not burden user with required minimum_alive_sec. If such field is not provided by the caller
+        # it will be set to the default value.
+        if minimum_alive_secs is None:
+            logger.info(f"client passed in empty minimum_alive_secs, using default value as:"
+                        f" {config.DEFAULT_TOKEN_MIN_ALIVE_TIME}")
+            minimum_alive_secs = config.DEFAULT_TOKEN_MIN_ALIVE_TIME
+        # Below code only need to deal with the case of cached entry could be found.
+        # The is, token_json could not be None
+        is_longer_than_minimum_alive_secs = satisfy_minimum_alive_secs(token_json, minimum_alive_secs)
+        if is_longer_than_minimum_alive_secs is True:
```

Comment:
I'm wondering if this logic chain and the one above can be cleaned up a little bit. Here's what I'm thinking in (very rough) psudocode

```
cached_token = dynamodb.get(client_id)
if cached_token is not None and cached_token is not expired:
    return cached_token

retrieve new token from Launchpad
store new token into dynamodb
return new token
```

Makes the logic flow a little easier to read/follow

Reply:
done